RecoPPS/Local/plugins/DiamondSampicDigiProducer.cc

```diff
@@ -0,0 +1,185 @@
+// -*- C++ -*-
+//
+// Package:    SampicDigi/DiamondSampicDigiProducer
+// Class:      DiamondSampicDigiProducer
+//
+/**\class DiamondSampicDigiProducer DiamondSampicDigiProducer.cc SampicDigi/DiamondSampicDigiProducer/plugins/DiamondSampicDigiProducer.cc
+
+ Description: This plugin takes testbeam data as input and converts them to TotemTimingDigi which could be then passed to reco
+
+ Implementation:
+     [Notes on implementation]
+*/
+//
+// Original Author:  Christopher Misan (krzysztof.misan@cern.ch)
+//         Created:  Sun, 07 Mar 2021 14:42:52 GMT
+//
+//
+
+// system include files
+#include <memory>
+#include <string>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/CTPPSDigi/interface/TotemTimingDigi.h"
+#include "DataFormats/CTPPSDetId/interface/TotemTimingDetId.h"
+#include "DataFormats/Common/interface/DetSetVector.h"
+#include "DataFormats/CTPPSDetId/interface/CTPPSDiamondDetId.h"
+
+#include "TFile.h"
+#include "TTree.h"
+#include "TChain.h"
+
+const int MAX_SAMPIC_CHANNELS = 32;
+const int SAMPIC_SAMPLES = 64;
+
+//
+// class declaration
+//
+
+class DiamondSampicDigiProducer : public edm::stream::EDProducer<> {
+public:
+  explicit DiamondSampicDigiProducer(const edm::ParameterSet&);
+  ~DiamondSampicDigiProducer() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  // ----------member data ---------------------------
+  std::vector<std::string> sampicFilesVec;
+  std::unordered_map<unsigned int, std::vector<unsigned int>> detid_vs_chid_;
+  TChain* inputTree;
+  std::stringstream ssLog;
+};
+
+//
+// constructors and destructor
+//
+
+using namespace edm;
+using namespace std;
+
+DiamondSampicDigiProducer::DiamondSampicDigiProducer(const edm::ParameterSet& iConfig)
+    : sampicFilesVec(iConfig.getParameter<std::vector<std::string>>("sampicFilesVec")) {
+  for (const auto& id_map : iConfig.getParameter<std::vector<edm::ParameterSet>>("idsMapping"))
+    detid_vs_chid_[id_map.getParameter<unsigned int>("treeChId")] = id_map.getParameter<vector<unsigned int>>("detId");
+
+  inputTree = new TChain("desy");
```

Comment:
consider using unique_ptr instead, to avoid managing the lifetime of inputTree yourself.